fix expression enclosed in parentheses

0x501D opened the pull request:
Expression enclosed in parentheses has no effect in return block.
From lua doc:
Any expression enclosed in parentheses always results in only one value.
Thus, (f(x,y,z)) is always a single value, even if f returns several values.
(The value of (f(x,y,z)) is the first value returned by f or nil if f does
not return any values.)

https://www.lua.org/manual/5.1/manual.html#2.5

This code will fail with gopher-lua:

```lua
-- example 1:
local function mult_ret()
   return 1, 2
end

local function enclosed_ret()
   return (mult_ret())
end

a,b = enclosed_ret()
assert(a == 1)
assert(b == nil) -- fails: b = 2

-- example 2:
local function enclosed_vararg(...) return (...) end
a,b,c=enclosed_vararg(1, 2, 3)
assert(a == 1 and b == nil and c == nil) -- fail: b = 2, c = 3

-- example 3:
local function enclosed_assign(...)
    local a,b = (...)
    return a,b
end
a,b,c=enclosed_assign(1, 2, 3)
assert(a == 1 and b == nil and c == nil) -- fail: b = 2, c = 3
```
Closes #363

Totktonada commented:
I know ano more case:

```lua
tarantool> function first(...) return (...) end
---
...

tarantool> first(1, 2, 3)
---
- 1
...
```

If it is not handled deliberately, please, clarify it in the commit messsage and the pull request description. If it is easy to do, let's do.

0x501D committed:
fix expression enclosed in parentheses

Expression enclosed in parentheses has no effect in return block.
From lua doc:
Any expression enclosed in parentheses always results in only one value.
Thus, (f(x,y,z)) is always a single value, even if f returns several values.
(The value of (f(x,y,z)) is the first value returned by f or nil if f does
not return any values.)

https://www.lua.org/manual/5.1/manual.html#2.5

This code examples will fail with gopher-lua:

-- example 1:
local function mult_ret()
   return 1, 2
end

local function enclosed_ret()
   return (mult_ret())
end

a,b = enclosed_ret()
assert(a == 1)
assert(b == nil) -- fails: b = 2

-- example 2:
local function enclosed_vararg(...) return (...) end
a,b,c=enclosed_vararg(1, 2, 3)
assert(a == 1 and b == nil and c == nil) -- fail: b = 2, c = 3

-- example 3:
local function enclosed_assign(...)
    local a,b = (...)
    return a,b
end
a,b,c=enclosed_assign(1, 2, 3)
assert(a == 1 and b == nil and c == nil) -- fail: b = 2, c = 3

Closes #363

Totktonada commented on compile.go:
```diff
@@ -114,7 +114,11 @@ func isVarArgReturnExpr(expr ast.Expr) bool {
 		return !ex.AdjustRet
 	case *ast.Comma3Expr:
-		return true
+		if ex.AdjustRet {
+			return false
+		} else {
+			return true
+		}
```
Nit: `return !ex.AdjustRet` (as above)?

0x501D replied:
Agree

0x501D replied:
Fixed

Totktonada commented on parse/parser.go.y:
```diff
@@ -396,6 +396,9 @@ prefixexp:
             $$ = $1
         } |
         '(' expr ')' {
+            if ex, ok := $2.(*ast.Comma3Expr); ok {
+                ex.AdjustRet = true
+            }
```
Just curious: can we get rid of `afunctioncall` above and set `AdjustRet` in one place (here)?

Maybe it is not worth to touch the grammar too much, because the testing is not thorough, so it is just due to an interest. I don't propose to change the code.

0x501D replied:
I tried this, but has no luck to get it work.

Totktonada replied:
Okay, thanks!

Totktonada reviewed (approved):
LGTM.

Totktonada commented:
NB: Created as part of https://github.com/tarantool/tt/issues/14.
